NetKet/Machine/activations.hpp

```diff
@@ -0,0 +1,192 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_ACTIVATIONS_HPP
+#define NETKET_ACTIVATIONS_HPP
+
+#include <Eigen/Dense>
+#include <complex>
+#include <iostream>
+#include <random>
+#include <vector>
+
+namespace netket {
+
+/**
+  Abstract class for Activations.
+*/
+class AbstractActivation {
+ public:
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+  virtual inline void operator()(const VectorType &Z, VectorType &A) = 0;
+
+  // Z is the layer output before applying nonlinear function
+  // A = nonlinearfunction(Z)
+  // F = dL/dA is the derivative of A wrt the output L = log(psi(v))
+  // G is the place to write the output i.e. G = dL/dZ = dL/dA * dA/dZ
+  virtual inline void ApplyJacobian(const VectorType &Z, const VectorType &A,
+                                    const VectorType &F, VectorType &G) = 0;
+  virtual ~AbstractActivation() {}
+};
+
+inline double lncosh(double x) {
+  const double xp = std::abs(x);
+  if (xp <= 12.) {
+    return std::log(std::cosh(xp));
+  } else {
+    const static double log2v = std::log(2.);
+    return xp - log2v;
+  }
+}
+
+// ln(cos(x)) for std::complex argument
+// the modulus is computed by means of the previously defined function
+// for real argument
+inline std::complex<double> lncosh(std::complex<double> x) {
+  const double xr = x.real();
+  const double xi = x.imag();
+
+  std::complex<double> res = lncosh(xr);
+  res += std::log(
+      std::complex<double>(std::cos(xi), std::tanh(xr) * std::sin(xi)));
+
+  return res;
+}
+
+class Identity : public AbstractActivation {
+ private:
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+ public:
+  // A = Z
+  inline void operator()(const VectorType &Z, VectorType &A) {
+    A.noalias() = Z;
+  }
+
+  // Apply the (derivative of activation function) matrix J to a vector F
+  // A = Z
+  // J = dA / dZ = I
+  // G = J * F = F
+  inline void ApplyJacobian(const VectorType & /*Z*/, const VectorType & /*A*/,
+                            const VectorType &F, VectorType &G) {
+    G.noalias() = F;
+  }
+};
+
+class Lncosh : public AbstractActivation {
+ private:
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+ public:
+  // A = Lncosh(Z)
+  inline void operator()(const VectorType &Z, VectorType &A) {
```

Comment:
You should add the [`override` specifier](https://en.cppreference.com/w/cpp/language/override) where applicable (i.e., `operator()` and `ApplyJacobian` here (and also for the other subclasses).